### ISSUE TITLE
Show/hide tokens

### ISSUE THREAD

**Issue by tschubotz** [gnosis/safe#298]
# What is this feature about? (1 sentence)
This feature allows users to show and hide specific tokens.

# Why is it needed? What is the value? For whom do we build it?
We show known tokens with a balance > 0 automatically already. However there are scam tokens out there that make random airdrops. A user should be able to hide those.
Before receiving a token, some users check if the app supports that token, so you can be sure that you can transfer it out again. (Yes we support theoretically all ERC20s, but users check the UI still). This was confirmed in a usability test that we made in the past. Users opened the token list and scrolled through it to check which tokens are supported and which ones aren't. 

Besides that, it's a good practice to allow users to manually show/hide tokens in case the automatic display fails.

# High-level overview of the feature
- User can access a list of available tokens from our backend.
- User can show / hide single tokens.
- User can search for token by name or symbol
- Manually shown/enabled tokens remain on the assets overview even if their balance goes to 0
- Manually hidden/disabled tokens remain hidden even if their balance is > 0.

**Comment by rmeissner** [comment]
- `User can access a list of available tokens from our backend.` - spam tokens are not part of this list
- While I see the advantage of hiding tokens, do we need to show 0 balance tokens in the first version?
- Should we not filter spam tokens on the backend and for advanced display of tokens defer to the web interface?

**Comment by Georgi87** [comment]
What is the benefit of allowing to show tokens with 0 balance? I would suggest to only have a blacklist and not a whitelist.

I would suggest the following:
- Automatically show tokens with balance > 0.
- Allow user to blacklist tokens with balance > 0.

We already filter spam tokens as we use the etherscan list to whitelist tokens, which excludes Spamtokens to my knowledge.

**Comment by sche** [comment]
Do we need to transfer tokens manual ordering feature like it is already implemented in the iOS version?

**Comment by Georgi87** [comment]
This is not important in my opinion. We could implement different sorting features like sorting by name or fiat token value.

**Comment by sche** [comment]
Do we want at some point in the future to add a possibility to add tokens manually?

**Comment by Georgi87** [comment]
I would disagree:

This feature is only interesting for teams, which release new tokens and are not yet at the step where they can add this token to Etherscan.

In case those teams want to see the balance in the app anyways, we would allow a checkbox "Show only whitelisted tokens", which can be checked by default. In case it is unchecked, we show all tokens incl. "spamtokens", which will also include the token the team would like to show. We index all tokens on our backend and have this information available.

**Comment by tschubotz** [comment]
> * `User can access a list of available tokens from our backend.` - spam tokens are not part of this list

yes, but they are part of the token balance list that the backend returns.

> * While I see the advantage of hiding tokens, do we need to show 0 balance tokens in the first version?

I thought the token management in the legacy app is quite nice, but you're right this is not required right now.

> * Should we not filter spam tokens on the backend and for advanced display of tokens defer to the web interface?

yes, I'll add it to #286 . 

> Allow user to blacklist tokens with balance > 0.

If we assume out token list is curated, I don't why this should be necessary :)


tldr;
I suggest the following then:
- Display all tokens with balance > 0 that are not scammy.
- Add any whitelisting/blacklisting/enabling/disabling logic later.

**Comment by tschubotz** [comment]
All this is a UX question and not really a technical one, right? I discussed this with Sasha and Kristina.
There's 2 more things to consider:

- There is a use case for enabling a token with balance > 0: Before you receive a token, you check if the app supports that token, so you can be sure that you can transfer it out again. (Yes we support theoretically all ERC20s, but users check the UI still)
- The above was confirmed in a usability test that we made in the past according to Kristina. Users opened the token list and scrolled through it to check which tokens are supported and which ones aren't.

**Comment by rmeissner** [comment]
For point 1: Since you cannot transfer out from the app anyways right I do not think this is required

I would simplify the mobile app in a lot of places and this feature (custom tokens) is available on the web version. Therefore I would not add this in the app. 

We could also have a FAQ with a list of supported tokens ... would do the same job then with less complexity.

Also with our current endpoints the list of supported tokens from the relay would not return all tokens we show balance for, so we would need to define this behaviour again.

**Comment by tschubotz** [comment]
> I would simplify the mobile app in a lot of places and this feature (custom tokens) is available on the web version. Therefore I would not add this in the app.

I did not mean custom tokens (i.e. adding a token by address). I rather mean showing/hiding tokens from our curated backend list.

> Also with our current endpoints the list of supported tokens from the relay would not return all tokens we show balance for, so we would need to define this behaviour again.

Don't we need to adapt our endpoint anyway to return only balances of curated tokens?

**Comment by rmeissner** [comment]
> I did not mean custom tokens
Sorry should be more specific, just used that to reference the token selection to get your "custom list of tokens". As said I do not think that this is really required in a first version. I know we should also look ahead and think about the future, so for this the issue (`Show / hide token`) is something to regard, but I would not have this in Phase 1 or 2.

> Don't we need to adapt our endpoint anyway to return only balances of curated tokens?
Yes, there is an issue for that open (https://github.com/gnosis/safe-transaction-service/issues/41). But this is not part of the web ui why do we need to make it part of the mobile app in the initial version?

**Comment by Georgi87** [comment]
> There is a use case for enabling a token with balance > 0: Before you receive a token, you check if the app supports that token, so you can be sure that you can transfer it out again. (Yes we support theoretically all ERC20s, but users check the UI still)

I think this would be more a use case to show a list of supported tokens instead of a whitelist. I agree with Richard that both (blacklist or list of supported tokens) is probably not required in the first version.

In my opinion, the best user experience would be to offer only the option to show only curated tokens or all tokens without any other options of whitelisting or blacklisting. Fortunately we index indeed all ERC20 tokens (like Etherscan does). To my knowledge we have the whitelist based on the 1000 curated tokens from Etherscan, which should cover the 99% use case. The 1% use case can be covered with the option to show all tokens. This might include some spam tokens but I don't see this as an issue. Indeed, I think it adds transparency as users should always know what they own. I got confused once using the web ui when I saw an incoming transfer of a token, which then didn't show up in the balance overview because it was not on our whitelist.

**Comment by tschubotz** [comment]
I'm not convinced, I still prefer the manual and explicit approach where a user can manually enable/disable tokens. Yes our token list is quite extensive, but I don't agree that it's a 1% use case to add other tokens. For that there have been too many requests to add tokens in the recent months: https://github.com/gnosis/safe-relay-service/issues?q=is%3Aissue+add+token

However, let's start somewhere. i.e. let's start with the full list from the backend in phase 1, as you suggested. 
I think it's not valuable to discuss this further right now. I added it to phase 2 so we can pick up the discussion again then. :)

**Comment by Georgi87** [comment]
Agree. One thing that could be evaluated is how to curate our whitelist better. Etherscan has all a much bigger list of curated tokens but exposes only the first 1000. Maybe we can find a way to get all of them.

**Comment by Uxio0** [comment]
> Agree. One thing that could be evaluated is how to curate our whitelist better. Etherscan has a much bigger list of curated tokens but exposes only the first 1000. Maybe we can find a way to get all of them.

They just have 987 tokens on the list, so maybe they don't have more curated tokens (I wouldn't think so if they have exactly 1000 tokens on the list). We also have another alternatives like https://tokens.kleros.io/ , for example

**Comment by koeppelmann** [comment]
FWIW - I found myself already too often in the situation that the wallet (including the Safe app) I used did not showed the token I cared for that I would definitely see it as a plus if I have the assurance that I can add a token manually if for some reason it does not appear on the list the wallet is using.

**Comment by Georgi87** [comment]
@Uxio0 it seems to me that this is mostly related to price information. The Etherscan list sorts token only based on market cap.  It might be the case that there is currently no price information for them available for more than 987 tokens. All tokens, which were requested to be added to the Safe whitelist, are actually approved by Etherscan (they have an Icon and metadata). So the actual curated list is much bigger but afaik not exposed via an API. If we can cover all of those I hope it would cover the 99% use case.

**Comment by tschubotz** [comment]
Closing this since we have the backend curated list for now. And also #357
